iroh-resolver/src/unixfs_builder.rs

```diff
@@ -197,74 +187,197 @@ impl File {
         self.name = Some(name.into());
         self
     }
 
     pub fn content_bytes<B: Into<Bytes>>(&mut self, content: B) -> &mut Self {
         let bytes = content.into();
-        self.content = Some(Box::pin(std::io::Cursor::new(bytes)));
+        self.reader = Some(Box::pin(std::io::Cursor::new(bytes)));
         self
     }
 
     pub fn content_reader<T: tokio::io::AsyncRead + 'static>(&mut self, content: T) -> &mut Self {
-        self.content = Some(Box::pin(content));
+        self.reader = Some(Box::pin(content));
         self
     }
 
     pub async fn build(self) -> Result<File> {
-        // encodes files as raw
+        if let Some(path) = self.path {
+            let name = match self.name {
+                Some(n) => n,
+                None => path
+                    .file_name()
+                    .and_then(|s| s.to_str())
+                    .unwrap_or_default()
+                    .to_string(),
+            };
+            return Ok(File {
+                content: Content::Path(path),
+                name,
+                chunker: Chunker::fixed_size(),
+                tree_builder: TreeBuilder::balanced_tree(),
+            });
+        }
+
+        if let Some(reader) = self.reader {
+            let name = self.name.ok_or_else(|| {
+                anyhow::anyhow!("must add a name when building a file from a reader or bytes")
+            })?;
+
+            return Ok(File {
+                content: Content::Reader(reader),
+                name,
+                chunker: Chunker::fixed_size(),
+                tree_builder: TreeBuilder::balanced_tree(),
+            });
+        }
+        anyhow::bail!("must have a path to the content or a reader for the content");
+    }
+}
+
+/// Entry is the kind of entry in a directory can be either a file or a
+/// folder (if recursive directories are allowed)
+#[derive(Debug, PartialEq)]
+enum Entry {
+    File(File),
+    Directory(Directory),
+}
+
+/// Construct a UnixFS directory.
+#[derive(Debug)]
+pub struct DirectoryBuilder {
+    name: Option<String>,
+    entries: Vec<Entry>,
+    typ: DirectoryType,
+    recursive: bool,
+}
+
+impl Default for DirectoryBuilder {
+    fn default() -> Self {
+        Self {
+            name: None,
+            entries: Default::default(),
+            typ: DirectoryType::Basic,
+            recursive: false,
+        }
+    }
+}
 
-        let name = self.name.ok_or_else(|| anyhow!("missing name"))?;
-        let reader = self.content.ok_or_else(|| anyhow!("missing content"))?;
+impl DirectoryBuilder {
+    pub fn new() -> Self {
+        Default::default()
+    }
+
+    pub fn as_hamt(&mut self) -> &mut Self {
+        self.typ = DirectoryType::Hamt;
+        self
+    }
+
+    pub fn allow_recursive(&mut self) -> &mut Self {
+        self.recursive = true;
+        self
+    }
+
+    pub fn with_name<N: Into<String>>(&mut self, name: N) -> &mut Self {
```

Comment:
why the change to `with_name` instead of `name`?

Reply:
We don't have to do this. I just personally find it confusing, expecting a method called `name` to return the content in the name field.

I've also seen `named` as a way to describe "this method is adding a name".

If you have a personal preference for `name`, or it's just the "rust-y" way to do it, I'm happy to change it back!

Reply:
I like consistency, so either all the builder methods should be prefixed with `with_` or `add_`, or we should drop it. up to you

Reply:
dropping it, if we are going for uniform.

Reply:
sorry, I didn't mean to remove the `add_`